src/lib/io/sol-uart-impl-linux.c

```diff
@@ -174,71 +321,60 @@ SOL_API void
 sol_uart_close(struct sol_uart *uart)
 {
     SOL_NULL_CHECK(uart);
-    if (uart->async.rx_fd_handler)
-        sol_fd_del(uart->async.rx_fd_handler);
-    if (uart->async.tx_fd_handler)
-        sol_fd_del(uart->async.tx_fd_handler);
-    close(uart->fd);
-    free(uart);
-}
+    SOL_EXP_CHECK(uart->delete_me);
 
-static void
-uart_tx_dispatch(struct sol_uart *uart, int status)
-{
-    uart->async.tx_fd_handler = NULL;
-    if (!uart->async.tx_cb)
+    if (uart->in_use) {
+        uart->delete_me = true;
         return;
-    uart->async.tx_cb((void *)uart->async.tx_user_data, uart,
-        (uint8_t *)uart->async.tx_buffer, status);
+    }
+    close_uart(uart);
 }
 
-static bool
-uart_tx_callback(void *data, int fd, uint32_t active_flags)
+SOL_API int
+sol_uart_write(struct sol_uart *uart, struct sol_blob *blob)
 {
-    struct sol_uart *uart = data;
-    int ret;
+    int r;
+    size_t total;
+    bool fd_created = false;
 
-    if (active_flags & FD_ERROR_FLAGS) {
-        SOL_ERR("Error flag was set on UART file descriptor %d.", fd);
-        ret = -1;
-        goto error;
-    }
+    SOL_NULL_CHECK(uart, -EINVAL);
+    SOL_NULL_CHECK(blob, -EINVAL);
+    SOL_EXP_CHECK(uart->delete_me, -EINVAL);
 
-    if (uart->async.tx_index == uart->async.tx_length) {
-        uart_tx_dispatch(uart, uart->async.tx_index);
-        return false;
-    }
+    r = sol_util_size_add(uart->pending_tx, blob->size, &total);
+    SOL_INT_CHECK(r, < 0, r);
 
-    ret = write(fd, uart->async.tx_buffer + uart->async.tx_index,
-        uart->async.tx_length - uart->async.tx_index);
-    if (ret < 0) {
-        SOL_ERR("Error when writing to file descriptor %d.", fd);
-        goto error;
-    }
+    if (uart->tx_size && total >= uart->tx_size)
```

Comment:
Small nitpick (feel free to ignore it): you could calculate total only if `uart->tx_size > 0`.